File: backend/src/services/ProxyService.ts
import { getDatabase } from '../database/connection';
import { ProxyEntry, ImportResult, StatsData } from '../types';
import { generateId, parseProxyUrl, isValidIp, isValidPort } from '../utils';

export class ProxyService {
  async addProxy(proxy: Omit<ProxyEntry, 'id' | 'createdAt'>): Promise<ProxyEntry> {
    const db = getDatabase();

    if (!isValidIp(proxy.host) && !/^[a-zA-Z0-9.-]+$/.test(proxy.host)) {
      throw new Error('无效的主机地址');
    }

    if (!isValidPort(proxy.port)) {
      throw new Error('无效的端口号');
    }

    const existing = await db.get(
      'SELECT id FROM proxies WHERE protocol = ? AND host = ? AND port = ?',
      proxy.protocol, proxy.host, proxy.port
    );

    if (existing) {
      throw new Error('代理已存在');
    }

    const id = generateId();
    const createdAt = new Date().toISOString();

    await db.run(
      `INSERT INTO proxies (id, protocol, host, port, username, password, country, isValid, createdAt)
       VALUES (?, ?, ?, ?, ?, ?, ?, 0, ?)`,
      [id, proxy.protocol, proxy.host, proxy.port, proxy.username || null, proxy.password || null, proxy.country || null, createdAt]
    );

    return { ...proxy, id, createdAt, isValid: false };
  }

  async addProxies(proxies: Omit<ProxyEntry, 'id' | 'createdAt'>[]): Promise<ImportResult> {
    const added: ProxyEntry[] = [];
    let duplicates = 0;

    for (const proxy of proxies) {
      try {
        const result = await this.addProxy(proxy);
        added.push(result);
      } catch (error: any) {
        if (error.message.includes('已存在')) {
          duplicates++;
        }
      }
    }

    return { added, duplicates };
  }

  async getAllProxies(): Promise<ProxyEntry[]> {
    const db = getDatabase();
    const rows = await db.all('SELECT * FROM proxies ORDER BY createdAt DESC');
    return rows.map(row => ({
      ...row,
      isValid: Boolean(row.isValid)
    }));
  }

  async getProxiesPaged(options: {
    page: number;
    pageSize: number;
    protocol?: string;
    status?: 'valid' | 'invalid' | 'all';
    maxResponseTime?: number;
    country?: string;
  }): Promise<{ data: ProxyEntry[]; total: number }> {
    const db = getDatabase();
    const { page, pageSize, protocol, status, maxResponseTime, country } = options;
    const conditions: string[] = [];
    const params: any[] = [];

    if (protocol && protocol !== 'all') {
      conditions.push('protocol = ?');
      params.push(protocol);
    }
    if (status === 'valid') {
      conditions.push('isValid = 1');
    } else if (status === 'invalid') {
      conditions.push('isValid = 0');
    }
    if (maxResponseTime) {
      conditions.push('responseTime IS NOT NULL AND responseTime <= ?');
      params.push(maxResponseTime);
    }
    if (country === 'unknown') {
      conditions.push('(country IS NULL OR country = \'\')' );
    } else if (country && country !== 'all') {
      conditions.push('country = ?');
      params.push(country);
    }

    const where = conditions.length > 0 ? 'WHERE ' + conditions.join(' AND ') : '';
    const countRow = await db.get(`SELECT COUNT(*) as total FROM proxies ${where}`, params);
    const total = countRow?.total ?? 0;

    const offset = (page - 1) * pageSize;
    const rows = await db.all(
      `SELECT * FROM proxies ${where} ORDER BY createdAt DESC LIMIT ? OFFSET ?`,
      [...params, pageSize, offset]
    );

    return {
      data: rows.map(row => ({ ...row, isValid: Boolean(row.isValid) })),
      total
    };
  }

  async getProxyById(id: string): Promise<ProxyEntry | null> {
    const db = getDatabase();
    const row = await db.get('SELECT * FROM proxies WHERE id = ?', id);
    if (!row) return null;
    return { ...row, isValid: Boolean(row.isValid) };
  }

  async getValidProxy(protocol?: string, maxResponseTime?: number): Promise<ProxyEntry | null> {
    const db = getDatabase();
    let query = 'SELECT * FROM proxies WHERE isValid = 1';
    const params: any[] = [];

    if (protocol) {
      query += ' AND protocol = ?';
      params.push(protocol);
    }

    if (maxResponseTime !== undefined) {
      query += ' AND responseTime <= ?';
      params.push(maxResponseTime);
    }

    query += ' ORDER BY RANDOM() LIMIT 1';

    const row = await db.get(query, params);
    if (!row) return null;
    return { ...row, isValid: Boolean(row.isValid) };
  }

  async getValidProxyCandidates(protocol?: string, maxResponseTime?: number, limit: number = 10, excludeIds: string[] = []): Promise<ProxyEntry[]> {
    const db = getDatabase();
    let query = 'SELECT * FROM proxies WHERE isValid = 1';
    const params: any[] = [];

    if (protocol) {
      query += ' AND protocol = ?';
      params.push(protocol);
    }

    if (maxResponseTime !== undefined) {
      query += ' AND responseTime <= ?';
      params.push(maxResponseTime);
    }

    if (excludeIds.length > 0) {
      query += ` AND id NOT IN (${excludeIds.map(() => '?').join(',')})`;
      params.push(...excludeIds);
    }

    query += ' ORDER BY RANDOM() LIMIT ?';
    params.push(limit);

    const rows = await db.all(query, params);
    return rows.map(row => ({ ...row, isValid: Boolean(row.isValid) }));
  }

  async deleteInvalidProxies(): Promise<number> {
    const db = getDatabase();
    const result = await db.get('SELECT COUNT(*) as count FROM proxies WHERE isValid = 0');
    const count = result?.count ?? 0;
    await db.run('DELETE FROM proxies WHERE isValid = 0');
    return count;
  }

  async deleteProxy(id: string): Promise<void> {
    const db = getDatabase();
    await db.run('DELETE FROM proxies WHERE id = ?', id);
  }

  async deleteProxies(ids: string[]): Promise<void> {
    const db = getDatabase();
    const placeholders = ids.map(() => '?').join(',');
    await db.run(`DELETE FROM proxies WHERE id IN (${placeholders})`, ids);
  }

  async updateProxyValidation(id: string, isValid: boolean, responseTime?: number): Promise<void> {
    const db = getDatabase();
    const lastChecked = new Date().toISOString();

    await db.run(
      'UPDATE proxies SET isValid = ?, lastChecked = ?, responseTime = ? WHERE id = ?',
      [isValid ? 1 : 0, lastChecked, responseTime || null, id]
    );
  }

  async updateProxyCountry(id: string, country: string): Promise<void> {
    const db = getDatabase();
    await db.run('UPDATE proxies SET country = ? WHERE id = ?', [country, id]);
  }

  async getCountries(): Promise<string[]> {
    const db = getDatabase();
    const rows = await db.all(
      `SELECT DISTINCT country FROM proxies WHERE country IS NOT NULL AND country != '' ORDER BY country`
    );
    return rows.map(r => r.country as string);
  }

  async getStats(): Promise<StatsData> {
    const db = getDatabase();

    const total = await db.get('SELECT COUNT(*) as count FROM proxies');
    const valid = await db.get('SELECT COUNT(*) as count FROM proxies WHERE isValid = 1');
    const byProtocol = await db.all(
      'SELECT protocol, COUNT(*) as total, SUM(isValid) as valid FROM proxies GROUP BY protocol'
    );

    const byProtocolMap: Record<string, { total: number; valid: number; invalid: number }> = {};
    byProtocol.forEach(row => {
      const rowTotal = row.total as number;
      const rowValid = (row.valid as number) || 0;
      byProtocolMap[row.protocol] = {
        total: rowTotal,
        valid: rowValid,
        invalid: rowTotal - rowValid
      };
    });

    return {
      total: total.count,
      valid: valid.count,
      invalid: total.count - valid.count,
      byProtocol: byProtocolMap
    };
  }

  async importFromText(text: string): Promise<ImportResult> {
    const lines = text.split('\n').map(line => line.trim()).filter(line => line);
    const proxies: Omit<ProxyEntry, 'id' | 'createdAt'>[] = [];

    for (const line of lines) {
      const parsed = parseProxyUrl(line);
      if (parsed) {
        proxies.push({
          protocol: parsed.protocol as any,
          host: parsed.host,
          port: parsed.port,
          username: parsed.username,
          password: parsed.password,
          isValid: false
        });
      }
    }

    return this.addProxies(proxies);
  }

  async exportToText(ids?: string[]): Promise<string> {
    const db = getDatabase();
    let query = 'SELECT * FROM proxies';
    let params: any[] = [];

    if (ids && ids.length > 0) {
      const placeholders = ids.map(() => '?').join(',');
      query += ` WHERE id IN (${placeholders})`;
      params = ids;
    }

    const rows = await db.all(query, params);

    return rows.map(row => {
      const auth = row.username && row.password ? `${row.username}:${row.password}@` : '';
      return `${row.protocol}://${auth}${row.host}:${row.port}`;
    }).join('\n');
  }
}

File: backend/services/proxy_service.py
from database import get_connection
from utils import generate_id, parse_proxy_url, is_valid_ip, is_valid_port
from datetime import datetime
from typing import List, Dict, Any, Optional, Tuple


class ProxyService:
    """代理服务"""
    
    async def add_proxy(self, proxy: Dict[str, Any]) -> Dict[str, Any]:
        """添加单个代理"""
        if not is_valid_ip(proxy['host']) and not re.match(r'^[a-zA-Z0-9.-]+$', proxy['host']):
            raise ValueError('无效的主机地址')
        
        if not is_valid_port(proxy['port']):
            raise ValueError('无效的端口号')
        
        with get_connection() as conn:
            cursor = conn.cursor()
            
            cursor.execute(
                'SELECT id FROM proxies WHERE protocol = ? AND host = ? AND port = ?',
                (proxy['protocol'], proxy['host'], proxy['port'])
            )
            
            if cursor.fetchone():
                raise ValueError('代理已存在')
            
            proxy_id = generate_id()
            created_at = datetime.now().isoformat()
            
            cursor.execute('''
                INSERT INTO proxies (id, protocol, host, port, username, password, country, isValid, createdAt)
                VALUES (?, ?, ?, ?, ?, ?, ?, 0, ?)
            ''', (proxy_id, proxy['protocol'], proxy['host'], proxy['port'],
                  proxy.get('username'), proxy.get('password'), proxy.get('country'), created_at))
            
            return {
                **proxy,
                'id': proxy_id,
                'createdAt': created_at,
                'isValid': False
            }
    
    async def add_proxies(self, proxies: List[Dict[str, Any]]) -> Dict[str, Any]:
        """批量添加代理"""
        added = []
        duplicates = 0
        
        for proxy in proxies:
            try:
                result = await self.add_proxy(proxy)
                added.append(result)
            except ValueError as e:
                if '已存在' in str(e):
                    duplicates += 1
        
        return {'added': added, 'duplicates': duplicates}
    
    async def get_all_proxies(self) -> List[Dict[str, Any]]:
        """获取所有代理"""
        with get_connection() as conn:
            cursor = conn.cursor()
            cursor.execute('SELECT * FROM proxies ORDER BY createdAt DESC')
            rows = cursor.fetchall()
            
            return [dict(row) for row in rows]
    
    async def get_proxies_paged(self, page: int, page_size: int, protocol: str = 'all',
                                status: str = 'all', max_response_time: Optional[int] = None,
                                country: str = 'all') -> Dict[str, Any]:
        """分页获取代理"""
        conditions = []
        params = []
        
        if protocol and protocol != 'all':
            conditions.append('protocol = ?')
            params.append(protocol)
        
        if status == 'valid':
            conditions.append('isValid = 1')
        elif status == 'invalid':
            conditions.append('isValid = 0')
        
        if max_response_time:
            conditions.append('responseTime IS NOT NULL AND responseTime <= ?')
            params.append(max_response_time)
        
        if country == 'ZZ':
            conditions.append('(country IS NULL OR country = \'\' OR country = \'ZZ\')')
        elif country and country != 'all':
            conditions.append('country = ?')
            params.append(country)
        
        where = ' AND '.join(conditions) if conditions else '1=1'
        
        with get_connection() as conn:
            cursor = conn.cursor()
            
            cursor.execute(f'SELECT COUNT(*) as total FROM proxies WHERE {where}', params)
            total = cursor.fetchone()['total']
            
            offset = (page - 1) * page_size
            cursor.execute(
                f'SELECT * FROM proxies WHERE {where} ORDER BY createdAt DESC LIMIT ? OFFSET ?',
                params + [page_size, offset]
            )
            rows = cursor.fetchall()
            
            return {
                'data': [dict(row) for row in rows],
                'total': total
            }
    
    async def get_valid_proxy(self, protocol: Optional[str] = None, max_response_time: Optional[int] = None,
                             count: int = 1, country: Optional[str] = None,
                             country_mode: str = 'include') -> List[Dict[str, Any]]:
        """获取有效代理"""
        query = 'SELECT * FROM proxies WHERE isValid = 1'
        params = []
        
        if protocol:
            query += ' AND protocol = ?'
            params.append(protocol)
        
        if max_response_time is not None:
            query += ' AND responseTime <= ?'
            params.append(max_response_time)
        
        if country:
            if country_mode == 'exclude':
                query += ' AND (country IS NULL OR country != ?)'
            else:
                query += ' AND country = ?'
            params.append(country)
        
        query += ' ORDER BY RANDOM() LIMIT ?'
        params.append(count)
        
        with get_connection() as conn:
            cursor = conn.cursor()
            cursor.execute(query, params)
            rows = cursor.fetchall()
            return [dict(row) for row in rows]
    
    async def get_valid_proxy_candidates(self, protocol: Optional[str] = None,
                                        max_response_time: Optional[int] = None,
                                        limit: int = 10, exclude_ids: List[str] = None,
                                        country_filter: Optional[str] = None,
                                        country_filter_mode: str = 'include') -> List[Dict[str, Any]]:
        """获取有效代理候选"""
        query = 'SELECT * FROM proxies WHERE isValid = 1'
        params = []
        
        if protocol:
            query += ' AND protocol = ?'
            params.append(protocol)
        
        if max_response_time is not None:
            query += ' AND responseTime <= ?'
            params.append(max_response_time)
        
        if country_filter:
            codes = [c.strip().upper() for c in country_filter.split(',') if c.strip()]
            if codes:
                placeholders = ','.join(['?' for _ in codes])
                if country_filter_mode == 'exclude':
                    query += f' AND (country IS NULL OR UPPER(country) NOT IN ({placeholders}))'
                else:
                    query += f' AND UPPER(country) IN ({placeholders})'
                params.extend(codes)
        
        if exclude_ids:
            placeholders = ','.join(['?' for _ in exclude_ids])
            query += f' AND id NOT IN ({placeholders})'
            params.extend(exclude_ids)
        
        query += ' ORDER BY RANDOM() LIMIT ?'
        params.append(limit)
        
        with get_connection() as conn:
            cursor = conn.cursor()
            cursor.execute(query, params)
            rows = cursor.fetchall()
            return [dict(row) for row in rows]
    
    async def delete_invalid_proxies(self) -> int:
        """删除无效代理"""
        with get_connection() as conn:
            cursor = conn.cursor()
            cursor.execute('SELECT COUNT(*) as count FROM proxies WHERE isValid = 0')
            count = cursor.fetchone()['count']
            cursor.execute('DELETE FROM proxies WHERE isValid = 0')
            return count
    
    async def delete_proxy(self, proxy_id: str) -> None:
        """删除单个代理"""
        with get_connection() as conn:
            cursor = conn.cursor()
            cursor.execute('DELETE FROM proxies WHERE id = ?', (proxy_id,))
    
    async def delete_proxies(self, ids: List[str]) -> None:
        """批量删除代理"""
        with get_connection() as conn:
            cursor = conn.cursor()
            placeholders = ','.join(['?' for _ in ids])
            cursor.execute(f'DELETE FROM proxies WHERE id IN ({placeholders})', ids)
    
    async def update_proxy_validation(self, proxy_id: str, is_valid: bool,
                                     response_time: Optional[int] = None) -> None:
        """更新代理验证状态"""
        last_checked = datetime.now().isoformat()
        
        with get_connection() as conn:
            cursor = conn.cursor()
            cursor.execute('''
                UPDATE proxies SET isValid = ?, lastChecked = ?, responseTime = ? WHERE id = ?
            ''', (1 if is_valid else 0, last_checked, response_time, proxy_id))
    
    async def update_proxy_country(self, proxy_id: str, country: str) -> None:
        """更新代理国家信息"""
        with get_connection() as conn:
            cursor = conn.cursor()
            cursor.execute('UPDATE proxies SET country = ? WHERE id = ?', (country, proxy_id))
    
    async def get_countries(self) -> List[str]:
        """获取所有国家代码"""
        with get_connection() as conn:
            cursor = conn.cursor()
            cursor.execute('''
                SELECT DISTINCT country FROM proxies
                WHERE country IS NOT NULL AND country != '' ORDER BY country
            ''')
            return [row['country'] for row in cursor.fetchall()]
    
    async def get_stats(self) -> Dict[str, Any]:
        """获取统计信息"""
        with get_connection() as conn:
            cursor = conn.cursor()
            
            cursor.execute('SELECT COUNT(*) as count FROM proxies')
            total = cursor.fetchone()['count']
            
            cursor.execute('SELECT COUNT(*) as count FROM proxies WHERE isValid = 1')
            valid = cursor.fetchone()['count']
            
            cursor.execute('''
                SELECT protocol, COUNT(*) as total, SUM(isValid) as valid
                FROM proxies GROUP BY protocol
            ''')
            by_protocol_rows = cursor.fetchall()
            
            by_protocol = {}
            for row in by_protocol_rows:
                row_valid = row['valid'] or 0
                by_protocol[row['protocol']] = {
                    'total': row['total'],
                    'valid': row_valid,
                    'invalid': row['total'] - row_valid
                }
            
            return {
                'total': total,
                'valid': valid,
                'invalid': total - valid,
                'byProtocol': by_protocol
            }
    
    async def import_from_text(self, text: str) -> Dict[str, Any]:
        """从文本导入代理"""
        lines = [line.strip() for line in text.split('\n') if line.strip()]
        proxies = []
        
        for line in lines:
            parsed = parse_proxy_url(line)
            if parsed:
                proxies.append({
                    'protocol': parsed['protocol'],
                    'host': parsed['host'],
                    'port': parsed['port'],
                    'username': parsed.get('username'),
                    'password': parsed.get('password'),
                    'isValid': False
                })
        
        return await self.add_proxies(proxies)
    
    async def export_to_text(self, ids: Optional[List[str]] = None) -> str:
        """导出代理为文本"""
        with get_connection() as conn:
            cursor = conn.cursor()
            
            if ids:
                placeholders = ','.join(['?' for _ in ids])
                cursor.execute(f'SELECT * FROM proxies WHERE id IN ({placeholders})', ids)
            else:
                cursor.execute('SELECT * FROM proxies')
            
            rows = cursor.fetchall()
            
            result = []
            for row in rows:
                auth = ''
                if row['username'] and row['password']:
                    auth = f"{row['username']}:{row['password']}@"
                result.append(f"{row['protocol']}://{auth}{row['host']}:{row['port']}")
            
            return '\n'.join(result)

import re
